From 969c3501e180f85b8ea2677426ec1548e1b3871d Mon Sep 17 00:00:00 2001
From: "Radu Pirea (NXP OSS)" <radu-nicolae.pirea@oss.nxp.com>
Date: Tue, 21 Sep 2021 12:01:08 +0300
Subject: [PATCH 17/19] s32g274aevb: probe sja1105 in misc_init_r

Probe sja1105 in misc_init_r in order to provide clock to the PFE2
interface.

Issue: ALB-4601

Upstream-Status: Pending 

Signed-off-by: Radu Pirea (NXP OSS) <radu-nicolae.pirea@oss.nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 arch/arm/cpu/armv8/s32/Kconfig         |  1 +
 arch/arm/dts/fsl-s32gxxxaevb.dtsi      |  7 +++++++
 board/freescale/s32-gen1/s32gxxxaevb.c | 20 ++++++++++++++++++++
 3 files changed, 28 insertions(+)

diff --git a/arch/arm/cpu/armv8/s32/Kconfig b/arch/arm/cpu/armv8/s32/Kconfig
index 343584739b..0512935f55 100644
--- a/arch/arm/cpu/armv8/s32/Kconfig
+++ b/arch/arm/cpu/armv8/s32/Kconfig
@@ -90,6 +90,7 @@ config TARGET_S32R45EMU
 config TARGET_S32G2XXAEVB
 	bool "Support s32g274aevb"
 	select NXP_S32G2XX
+	select MISC_INIT_R if SJA1105
 
 config TARGET_S32G274ARDB
 	bool "Support s32g274ardb"
diff --git a/arch/arm/dts/fsl-s32gxxxaevb.dtsi b/arch/arm/dts/fsl-s32gxxxaevb.dtsi
index 10201eb20b..65679d8dcb 100644
--- a/arch/arm/dts/fsl-s32gxxxaevb.dtsi
+++ b/arch/arm/dts/fsl-s32gxxxaevb.dtsi
@@ -45,6 +45,13 @@
 	pinctrl-0 = <&pinctrl0_dspi5 &pinctrl1_dspi5>;
 	pinctrl-names = "default";
 	status = "okay";
+
+	ethernet-switch@0 {
+		compatible = "nxp,sja1105-fw-loader";
+		spi-max-frequency = <1000000>;
+		reg = <0>;
+		spi-cpha;
+	};
 };
 
 &gmac0 {
diff --git a/board/freescale/s32-gen1/s32gxxxaevb.c b/board/freescale/s32-gen1/s32gxxxaevb.c
index 02805991c3..17845529a8 100644
--- a/board/freescale/s32-gen1/s32gxxxaevb.c
+++ b/board/freescale/s32-gen1/s32gxxxaevb.c
@@ -5,8 +5,11 @@
 #include <asm/arch/soc.h>
 #include <board_common.h>
 #include <common.h>
+#include <dm/uclass.h>
 #include <s32g274a_common.h>
 
+#define SJA1105_NAME	"ethernet-switch@0"
+
 void setup_iomux_uart(void)
 {
 #if (CONFIG_FSL_LINFLEX_MODULE == 0)
@@ -33,3 +36,20 @@ void setup_iomux_uart(void)
 #endif
 }
 
+int misc_init_r(void)
+{
+#if CONFIG_IS_ENABLED(NET) && CONFIG_IS_ENABLED(FSL_PFENG) && \
+	CONFIG_IS_ENABLED(SJA1105)
+	struct udevice *dev;
+
+	/* Probe sja1105 in order to provide a clock for the PFE2 interface,
+	 * otherwise clock init for this interface will fail.
+	 * The return value is not checked on purpose. If the S32G PROCEVB
+	 * board is used without PLATEVB board the uclass_get_device_by_name
+	 * call will fail and returning and error code will break the init_call
+	 * sequence of the u-boot.
+	 */
+	uclass_get_device_by_name(UCLASS_MISC, SJA1105_NAME, &dev);
+#endif
+	return 0;
+}
-- 
2.17.1

